### PR TITLE
chassis: Fix memory leak in vkCreateInstance if an error occured

### LINKED ISSUE #5076: Memory leaks when call down vkCreateInstance fails or other Vulkan API calls fail inside of vkCreateInstance
**Describe the Issue**
There are two leaks happening, one causing VVL to leak memory directly, and another causing memory leaks in the loader.

1. When VK_ERROR_OUT_OF_HOST_MEMORY is returned from the call to vkEnumeratePhysicalDevices that happens inside of VVL's `vkCreateInstance`, the validation layers cause the loader to leak memory. These Vulkan API calls are made inside `PreCallRecordCreateInstance` so require modification to the generated code. The problem is that when OOM is returned from `vkEnumeratePhysicalDevices`, the loader has successfully completed the vkCreateInstance terminator. OOM from one of the subsequent Vulkan API calls requires that VVL cleanup all vulkan objects before existing, which should only require calling vkDestroyInstance from inside VVL.

A description of the sequence of events
* Loader's vkCreateInstance (trampoline), calls into layer's vkCreateInstance
* VVL's vkCreateInstance is called, initializing local objects and setting up dispatch tables.
* VVL calls down vkCreateInstance
* Loader's terminator_CreateInstance is called and completes successfully
* Control flow returns to VVL, continues setting up objects
* `PreCallRecordCreateInstance` is called on the validation layer objects
* A vulkan API call is made, such as `vkEnumeratePhysicalDevices`, which fails with OOHM (error_out_of_host_memory)
* What happens now is: Ignore the error condition and continue as if nothing happened. 

What should happen is that VVL should catch that OOHM, delete the instance by calling vkDestroyInstance, then return immediately.

2. The structures directly leaked by VVL are the validation layer objects allocated here.
https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/d5426ea20669a4d8de34dd3b89685fba73a21c4f/layers/generated/chassis.cpp#L290

The solution would be to put them in std::unique_ptr<> wrappers immediately instead of having naked pointers laying about.

### PR DESCRIPTION
**Pull request by charles-lunarg**
If an error occurs duing the call down the CreateInstance chain, VVL must
clean up any allocated objects. This is made easier by only allocating
Validation Objects for which the filters allow, rather than the previous
method of allocating everything and removing the unused ones. While from a
first glance std::unique_ptr<> seems like the obvious choice, because of
how intrusive such a change would be it was not chosen.

The second place an error can occur is during PostCallRecordCreateInstance.
This results in faileding a vulkan function call after a successful call
down the vkCreateInstance chain, leaving Vulkan in a weird state. The
previous commit would not pass the error code upwards, so to the application
everything appeared correct since the return value was VK_SUCCESS. However,
this causes CTS Out Of Memory tests to fail due to OOM not being returned.
The fix for this is to make the PostCallRecordCreateInstance pass the
VkResult value by reference instead of value, allowing the error code to
propagate upwards. This did require special casing the PostCallRecord
in the various code generators.

Lastly, the commit does a general cleanup of the ValidationObject code by
removing the functions only used during initialization. Since they are used
only during vkCreateInstance and vkCreateDevice, moving them into those
functions cleans up the overal interface of the functions.

This commit makes the following tests have clean output when run with a
leak sanitizer:
* VkLayerTest.InstanceExtensionDependencies
* VkLayerTest.InstanceBadStype
* VkLayerTest.InstanceDuplicatePnextStype
* VkLayerTest.InstanceAppInfoBadStype
* VkLayerTest.InstanceValidationFeaturesBadFlags
* VkLayerTest.InstanceBadValidationFlags

Note that to get the clean output, the loader must have been updated to
contain the commit 3db19f3 (or just run with a loader built against
1.3.243 or higher).

Fixes #5076

### CODE REVIEW

File: layers/best_practices/best_practices_validation.h [hunk 1]
```diff
@@ -904,10 +904,6 @@ class BestPractices : public ValidationStateTracker {
         return std::static_pointer_cast<SWAPCHAIN_NODE>(std::make_shared<bp_state::Swapchain>(this, create_info, swapchain));
     }
 
-    std::shared_ptr<PHYSICAL_DEVICE_STATE> CreatePhysicalDeviceState(VkPhysicalDevice phys_dev) final {
```
Comment (ncesario-lunarg) [inline layers/best_practices/best_practices_validation.h:907]: Why was this removed? If I understand correctly, best practices will not get the correct derived state without this function.

File: layers/state_tracker/state_tracker.cpp [hunk 2]
```diff
@@ -4048,7 +4044,7 @@ void ValidationStateTracker::PostCallRecordCreateInstance(const VkInstanceCreate
     }
 
     for (auto physdev : physdev_handles) {
-        Add(CreatePhysicalDeviceState(physdev));
+        Add(std::make_shared<PHYSICAL_DEVICE_STATE>(physdev));
```
Comment (ncesario-lunarg) [inline layers/state_tracker/state_tracker.cpp:4047]: I might be missing something, but I don't think this is correct. See comment above.

File: layers/best_practices/best_practices_validation.h (hunk 1 above)
Comment (charles-lunarg) [reply layers/best_practices/best_practices_validation.h:907]: I misunderstood the code, yes this shouldn't be removed. I did this before realizing that the `ValidationStateTracker` object is a base class for others, rather than a leaf object.

File: layers/state_tracker/state_tracker.cpp (hunk 2 above)
Comment (charles-lunarg) [reply layers/state_tracker/state_tracker.cpp:4047]: Yep, will do the same here.

**Comment by charles-lunarg** [comment]
Okay it appears I can't re-request a review from both reviewers. Apologies for the spam.

Regardless, the code now passes tests. I went ahead and ran the code with leak-sanitizer and found several test cases which triggered a different leak, whose fix was also added.

Additionally, I fixed a couple of very low hanging fruit memory leaks, let me know if those should be in their own PR.

File: layers/generated/chassis.cpp [hunk 3]
```diff
@@ -312,16 +326,32 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateInstance(const VkInstanceCreateInfo *pCreat
...
         auto lock = intercept->ReadLock();
         skip |= intercept->PreCallValidateCreateInstance(pCreateInfo, pAllocator, pInstance);
-        if (skip) return VK_ERROR_VALIDATION_FAILED_EXT;
+        if (skip) {
+            for (ValidationObject* object : local_object_dispatch) {
+                delete object;
+            }
+            DeactivateInstanceDebugCallbacks(report_data);
+            FreePnextChain(report_data->instance_pnext_chain);
+            LayerDebugUtilsDestroyInstance(report_data);
```
Comment (juan-lunarg) [inline layers/generated/chassis.cpp:335]: Code duplication issue regarding freeing memory. Ideally this would be a function.

**Comment by charles-lunarg** [comment]
I am on the fence about this PR.
Its... big, and not because it has to be. There are 3 separate things being fixed here, VkCreateInstance leaking when errors occur, vkCreateInstance not propagating errors up from PostCallRecordCreateInstance, and misc memory leak fixes in some tests. I'm of the mind of splitting this into three separate PR's, and I'll probably do it 'right' as well, since I'm not sold on the fix for PostCallRecordCreateInstance.

So, I will close this PR and submit smaller ones to replace it.
